filters/include/pcl/filters/experimental/grid_filter.h

```diff
@@ -0,0 +1,230 @@
+    return save_leaf_layout_;
+  }
+
+  /** \brief Provide the name of the field to be used for filtering data. In conjunction
+   * with \a setFilterLimits, points having values outside this interval will be
+   * discarded.
+   * \param[in] field_name the name of the field that contains values used for filtering
+   */
+  inline void
+  setFilterFieldName(const std::string& field_name)
+  {
+    filter_field_name_ = field_name;
+  }
+
+  /** \brief Get the name of the field used for filtering. */
+  inline std::string const
+  getFilterFieldName() const
+  {
+    return filter_field_name_;
+  }
+
+  /** \brief Get the field filter limits (min/max) set by the user. The default values
+   * are numeric_limits<double>::min(), numeric_limits<double>::max().
+   * \param[out] limit_min the minimum allowed field value
+   * \param[out] limit_max the maximum allowed field value
+   */
+  inline void
+  getFilterLimits(double& limit_min, double& limit_max) const
+  {
+    limit_min = filter_limit_min_;
+    limit_max = filter_limit_max_;
+  }
+
+  /** \brief Set to true if we want to return the data outside the interval specified by
+   * setFilterLimits (min, max).
+   * Default: false.
+   * \param[in] limit_negative return data inside the interval (false) or outside (true)
+   */
+  inline void
+  setFilterLimitsNegative(const bool limit_negative)
+  {
+    filter_limit_negative_ = limit_negative;
+  }
+
+  /** \brief Get whether the data outside the interval (min/max) is to be returned
+   * (true) or inside (false).
+   * \param[out] limit_negative true if data \b outside the interval [min; max] is to be
+   * returned, false otherwise
+   */
+  inline void
+  getFilterLimitsNegative(bool& limit_negative) const
+  {
+    limit_negative = filter_limit_negative_;
+  }
+
+  /** \brief Get whether the data outside the interval (min/max) is to be returned
+   * (true) or inside (false).
+   * \return true if data \b outside the interval [min; max] is to be returned, false
+   * otherwise
+   */
+  inline bool
+  getFilterLimitsNegative() const
+  {
+    return filter_limit_negative_;
+  }
+
+protected:
+  /** \brief Set to true if all fields need to be downsampled, or false if just XYZ. */
+  bool downsample_all_data_;
+
+  /** \brief Set to true if leaf layout information needs to be saved in \a
+   * leaf_layout_. */
+  bool save_leaf_layout_;
+
+  /** \brief The desired user filter field name. */
+  std::string filter_field_name_;
+
+  /** \brief The minimum allowed filter value a point will be considered from. */
+  double filter_limit_min_;
+
+  /** \brief The maximum allowed filter value a point will be considered from. */
+  double filter_limit_max_;
+
+  /** \brief Set to true if we want to return the data outside (\a filter_limit_min_;\a
+   * filter_limit_max_).
+   * Default: false. */
+  bool filter_limit_negative_;
+
+  /** \brief Minimum number of points per voxel for the centroid to be computed */
+  std::size_t min_points_per_voxel_;
```

Comment:
This is the API that makes GridFilter non-minimalistic. By keeping this in GridFilter and moving the applyFilter and friends out into TransforAndFilter, reasoning about the required API for the filter becomes easier

Reply:
I see, I can try that out